src/components/select/multi-select/multi-select.component.js

```diff
@@ -111,6 +116,34 @@ const MultiSelect = React.forwardRef(
       [name, id]
     );
 
+    /* generic value update function which can be used for both controlled and uncontrolled
+     * components, both with and without onChange.
+     * - if onChange is provided, calls that with the updated value, provided it has changed in length
+     * (ie if an option has been selected or deselected)
+     * - if the component is uncontrolled, call setSelectedValue with the provided update function
+     * - if the component is controlled, do not change selectedValue and instead directly call
+     * the update function and pass its result to onChange
+     * - ensure that in every circumstance, onChange and the update function are only called at most
+     * once, in case they have side effects */
+    const setValue = useCallback(
+      (updateFunction) => {
+        const updateAndCallOnChange = (oldValue) => {
```

Comment:
In an ideal world I agree with you, but I think this needs to do both because of all the situations it's used in. I'll have another think about it though when I come back to this PR!